pkg/spdk/disk/nvme/nvme.go

```diff
@@ -0,0 +1,97 @@
+package nvme
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	commonTypes "github.com/longhorn/go-common-libs/types"
+	"github.com/longhorn/go-spdk-helper/pkg/jsonrpc"
+	spdkclient "github.com/longhorn/go-spdk-helper/pkg/spdk/client"
+	spdksetup "github.com/longhorn/go-spdk-helper/pkg/spdk/setup"
+	spdktypes "github.com/longhorn/go-spdk-helper/pkg/spdk/types"
+	helpertypes "github.com/longhorn/go-spdk-helper/pkg/types"
+	helperutil "github.com/longhorn/go-spdk-helper/pkg/util"
+
+	"github.com/longhorn/longhorn-spdk-engine/pkg/spdk/disk"
+)
+
+type DiskDriverNvme struct {
+}
+
+const (
+	diskDriver = "nvme"
+)
+
+func init() {
+	driver := &DiskDriverNvme{}
+	if err := disk.RegisterDiskDriver(diskDriver, driver); err != nil {
+		panic(err)
+	}
+}
+
+func (d *DiskDriverNvme) DiskCreate(spdkClient *spdkclient.Client, diskName, diskPath string, blockSize uint64) (string, error) {
+	// TODO: validate the diskPath
+	executor, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get the executor for NVMe disk create %v", diskPath)
+	}
+
+	_, err = spdksetup.Bind(diskPath, disk.UioDriverUioPciGeneric, executor)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to bind NVMe disk %v with %v", diskPath, disk.UioDriverUioPciGeneric)
+	}
+
+	bdevs, err := spdkClient.BdevNvmeAttachController(diskName, "", diskPath, "", "PCIe", "",
+		helpertypes.DefaultCtrlrLossTimeoutSec, helpertypes.DefaultReconnectDelaySec, helpertypes.DefaultFastIOFailTimeoutSec)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to attach NVMe disk %v", diskPath)
+	}
+	if len(bdevs) == 0 {
+		return "", errors.Errorf("failed to attach NVMe disk %v with %v", diskPath, diskDriver)
+	}
+	return bdevs[0], nil
+}
+
+func (d *DiskDriverNvme) DiskDelete(spdkClient *spdkclient.Client, diskName, diskPath string) (deleted bool, err error) {
+	executor, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get the executor for NVMe disk %v deletion", diskName)
+	}
+
+	_, err = spdkClient.BdevNvmeDetachController(diskName)
+	if err != nil {
+		if !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+			return false, errors.Wrapf(err, "failed to detach NVMe disk %v", diskName)
+		}
+	}
+
+	_, err = spdksetup.Unbind(diskPath, executor)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to unbind NVMe disk %v", diskPath)
+	}
+	return true, nil
+}
+
+func (d *DiskDriverNvme) DiskGet(spdkClient *spdkclient.Client, diskName, diskPath string, timeout uint64) ([]spdktypes.BdevInfo, error) {
+	bdevs, err := spdkClient.BdevGetBdevs("", 0)
```

Comment:
Why not use `bdev_nvme_get_controllers`?

Reply:
We need `bdev` rather than `controller` here.